### ISSUE TITLE
Change return type hint of `JsonCast::get` method to `mixed`

### ISSUE THREAD

**Issue by ChibuezeAgwuDennis** [codeigniter4/CodeIgniter4#9323]
## Description

   The `JsonCast::get` method currently has a return type hint of `array|stdClass`. However, considering the nature of the method, which decodes JSON strings into PHP types, the return value can theoretically be any PHP type, including scalar values, arrays, or `stdClass` objects.

## Suggested Change

   Change the return type hint of the `get` method from `array|stdClass` to `mixed`, to accurately reflect the potential return types.

## Code Example

   ```php
   public static function get(
       mixed $value,
       array $params = [],
       ?object $helper = null
   ): mixed {
       // Existing logic...
   }
   ```

## Rationale

   This change will prevent any type mismatches and make the method's expected behavior clearer to developers using the framework.

## Additional Context

   If the method is used with a JSON string representing a scalar value (like a string or number), the current type hints may cause confusion or type errors.

## References

   Relevant section in the codebase:
   ```php
   class JsonCast extends BaseCast {
       //...
   }
   ```

  Thank you for considering this improvement.

**Comment by neznaika0** [comment]
Tests for mixed values?
The normal behavior is an `array` or `stdClass`. Maybe string:
`var_dump(json_encode(null)); // "null"`

**Comment by ChibuezeAgwuDennis** [comment]
### Response to "Tests for Mixed Values?"

The suggestion to use `mixed` as the return type for the `JsonCast::get` method is meant to capture the full range of potential return types that can result from decoding a JSON string. This includes not only arrays and `stdClass` objects but also scalar values (like strings and numbers) and even `null`. Here's how to address this:

1. **Explain the Justification:**
   - The `json_decode` function in PHP indeed returns various types based on the JSON input. By setting the return type to `mixed`, the method signature would accurately reflect all possible outcomes of decoding JSON strings, including edge cases where the JSON might represent a scalar or `null`.

2. **Address the Current Use Cases:**
   - While the common cases involve arrays or `stdClass` objects, there are legitimate scenarios where JSON might represent a simple value or `null`. For example:
     ```php
     // JSON string representing a number
     $jsonNumber = "42";

     // JSON string representing a string
     $jsonString = "\"Hello, World\"";

     // JSON string representing null
     $jsonNull = "null";
     ```

3. **Propose Testing Strategy:**
   - To ensure that the method handles all possible return types correctly, below is how I implemented a series of tests aligned with CodeIgniter 4's testing standards. These tests covers:
     - **Array and `stdClass` objects:** to ensure that JSON objects and arrays are correctly decoded.
     - **Scalar Values:** to test cases for JSON strings representing numbers, strings, and booleans.
     - **Null Values:** to verify that JSON strings representing `null` are handled without errors.
     - **Invalid JSON:** to ensure that exceptions are thrown for invalid JSON strings.

4. **Illustrate Test Cases:**
   ```php
   public function testJsonCastGetWithVariousJsonTypes()
   {
       // Test for array
       $jsonArray = '[{"key": "value"}]';
       $this->assertIsArray(JsonCast::get($jsonArray, ['array']));

       // Test for stdClass
       $jsonObject = '{"key": "value"}';
       $this->assertInstanceOf(stdClass::class, JsonCast::get($jsonObject));

       // Test for scalar value (number)
       $jsonNumber = '42';
       $this->assertIsInt(JsonCast::get($jsonNumber));

       // Test for scalar value (string)
       $jsonString = '"Hello, World"';
       $this->assertIsString(JsonCast::get($jsonString));

       // Test for null
       $jsonNull = 'null';
       $this->assertNull(JsonCast::get($jsonNull));

       // Test for invalid JSON
       $this->expectException(CastException::class);
       JsonCast::get('invalid json');
   }
   ```

In summary, by adopting `mixed` as the return type and implementing comprehensive tests, we can ensure that the `JsonCast::get` method is robust, flexible, accommodating the diverse nature of JSON data. This approach not only addresses potential type mismatches but also enriches the framework's reliability and developer experience.

**Comment by neznaika0** [comment]
`int|null|bool|float` It is not returned

**Comment by paulbalandan** [comment]
While `json_decode` may return scalars in some cases, do we have valid use cases for such inputs that would cause such returns within the context of CI4?

By virtue of YAGNI, if the use case for widening the return is because the underlying implementation theoretically covers such wide return but realistically no one ever enter into such scenario, I think it is not worth it.

**Comment by ChibuezeAgwuDennis** [comment]
### Reply to @neznaika:

You raised a valid point regarding the specific types `int|null|bool|float` not being explicitly returned by the current implementation. However, the intention behind proposing the use of `mixed` as the return type is to future-proof the method against any potential scenarios where these types might be encountered.

By leveraging `mixed`, we ensure that the method can handle any valid JSON input without any type-related constraints, thus enhancing its robustness. While the current use cases primarily involve arrays and `stdClass` objects, it’s prudent to consider the full spectrum of JSON data types.

### Reply to @paulbalandan:

I appreciate your thoughtful consideration of the YAGNI principle, and I agree that practicality should guide our decisions. However, it’s worth noting that JSON data, especially when dealing with external APIs or user-generated content, can often be unpredictable. The use of `mixed` as a return type is a precaution to handle any eventuality gracefully.

While it might seem that scalar returns are not common within the current context of CI4, one could envision scenarios such as settings storage or configuration management where JSON-encoded scalar values are not only plausible but also practical. For example:

#### Scenarios in a CodeIgniter 4 Model Cast for Key-Value Options:

1. **Configuration Settings:**
   - A model that stores various configuration settings where each setting might be a different type, such as a boolean for feature toggles, a float for precision settings, or an integer for limits.

   ```php
   class SettingsModel extends \CodeIgniter\Model
   {
       protected $table = 'settings';
       protected $casts = [
           'value' => 'json',
       ];
   }

   // Example JSON values:
   // {"feature_enabled": true}
   // {"timeout": 30}
   // {"version": "1.2.3"}
   ```

2. **User Preferences:**
   - A scenario where user preferences are stored in a JSON format, allowing for varied data types like strings, integers, and booleans to represent different settings.

   ```php
   // Saving user preferences
   $preferences = [
       'theme' => 'dark',
       'notifications' => true,
       'font_size' => 14
   ];
   
   $userModel->update($userId, ['preferences' => json_encode($preferences)]);
   ```

3. **Dynamic Form Data:**
   - For applications that use dynamic forms, the data might be stored in JSON, where each form element can be a different type, such as a dropdown value (string), a checkbox (boolean), or a slider (integer).

   ```php
   // Example JSON structure:
   // {"dropdown": "option1", "checkbox": false, "slider_value": 25}
   ```

The current framework usage might not heavily rely on scalar JSON types, adopting a `mixed` return type could provide a flexible foundation for future developments and integrations. This approach ensures that the framework remains adaptable to a broader range of use cases, safeguarding against unforeseen needs without introducing significant overhead.

**Comment by paulbalandan** [comment]
You are not giving a valid use case for scenarios where inputs are purely scalar. In your example 1, the JSON values are actually returned as json objects (or json arrays if required).

```php
$json = '{"featured_enabled": true}';

// this is an object
$value = JsonCast::get($json);
var_dump($value->feature_enabled); // bool(true)
```

**Comment by paulbalandan** [comment]
Please stop using ChatGPT for giving us examples. Use a redacted version of your real world projects (or needs) where such scenarios to return `null|bool` is needed to be returned.

**Comment by ChibuezeAgwuDennis** [comment]
here is the examples

Array (
          [id] => 9 
          [key] => site_name
          [value] => "Sarovus CMS" this was JsonCast::set(...) which is a string
          [type] => string 
          [group] => general_settings 
          [created_at] => 2024-12-15 04:50:23 
          [updated_at] => 2024-12-15 04:50:23
)

**Comment by paulbalandan** [comment]
Hi @ChibuezeAgwuDennis I do not mean to offend you in any way. But to me or maybe other readers that your previous responses are made entirely by chatgpt. I understand if you're not fluent in English and want the AI to help you construct your messages but the way I see it you type in the question and let the AI answer for you and that's it. No changes or anything.

**Comment by paulbalandan** [comment]
Now, my question on your example. Why is the cast you set for `value` is `json` when it is just a string?

**Comment by ChibuezeAgwuDennis** [comment]
@paulbalandan 
## Example in Model:

```php
<?php

namespace App\Models;

use Codeigniter\Model;

class Setting extends Model
{
    protected $table = 'settings';
    
    protected $allowedFields = [
        'key',
        'value',
        'type',
    ];
    
    protected array $casts = [
        'value' => 'json',
    ];
}
```

from the example above, We want to allows the `value` field to accommodate various data formats, including arrays, objects, and scalar types such as strings and numbers in order to ensure diverse types of data which can be efficiently stored and retrieved from the `value` column in the `settings` table. 

## My modification on the Core Code:

```php

public static function get(
        mixed $value,
        array $params = [],
        ?object $helper = null
    ): mixed {
        if (! is_string($value)) {
            self::invalidTypeValueError($value);
        }

        $associative = in_array('array', $params, true);

        $output = ($associative ? [] : new stdClass());

        try {
            $output = json_decode($value, $associative, 512, JSON_THROW_ON_ERROR);
        } catch (JsonException $e) {
            throw CastException::forInvalidJsonFormat($e->getCode());
        }

        return $output;
    }
```

I changed the return hint type to ``mixed`` and it works fine for any data types with experiencing any data type return error.

I am sorry if i have stressed many readers but just simple understanding i need. You are using json_encode and json_decode and not permitting all it's features on ``->get()``. 
You allowed all data types on setting json cast here

```php
public static function set(
        mixed $value, // any data type can be passed 
        array $params = [],
        ?object $helper = null
    ): string
```
While you are not allowing those data types return except ``array`` or ``object``. 

So check the code very well.
I don't want to touch the core code so that is the reason why i post my suggestions here.

**Comment by paulbalandan** [comment]
As no one has acted on a PR to do this, it seems the use case is limited. Closing as very stale. Please send a PR if request is still valid.
